crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1151,13 +1154,23 @@ impl<'db> TypeInferenceBuilder<'db> {
         } = assignment;
 
         for target in targets {
-            if let ast::Expr::Name(name) = target {
-                self.infer_definition(name);
-            } else {
-                // TODO infer definitions in unpacking assignment. When we do, this duplication of
-                // the "get `Expression`, call `infer_expression_types` on it, `self.extend`" dance
-                // will be removed; it'll all happen in `infer_assignment_definition` instead.
-                let expression = self.index.expression(value.as_ref());
+            self.infer_assignment_target(target, value);
+        }
+    }
+
+    // TODO: Remove the `value` argument once we handle all possible assignment targets.
+    fn infer_assignment_target(&mut self, target: &ast::Expr, value: &ast::Expr) {
+        match target {
+            ast::Expr::Name(name) => self.infer_definition(name),
+            ast::Expr::List(ast::ExprList { elts, .. })
+            | ast::Expr::Tuple(ast::ExprTuple { elts, .. }) => {
+                for element in elts {
+                    self.infer_assignment_target(element, value);
+                }
+            }
+            _ => {
+                // TODO: Remove this once we handle all possible assignment targets.
```

Comment:
Yeah, I think that makes sense.

I'm just trying to understand for "assignment that doesn't create a Definition." do you mean that something like `a.b = 1` won't create a `Definition` or it won't create an `AssignmentDefinition`?

Reply:
I mean that `a.b = 1` (and `a[b] = 1`) won't create a `Definition` at all, since it doesn't define any name in the local scope. Though ultimately it may create some other kind of struct that we track relative to control-flow, if we want to narrow types based on such assignments.

Reply:
It's not important that we do anything about this comment in this PR, the existing TODO is OK, we have more to figure out here regardless.